### DISCUSSION TITLE
android版本运行时仍然查询本地dns

### DISCUSSION THREAD

**Discussion by Waopes** [v2fly/v2ray-core#692] [Q&A] [answered]
根据白话文教程，成功编译Android arm64版本的应用程序后，在手机（华为P10，Android 9.0 & MI10，Android 11）上运行时仍然出现查询本地53端口的记录，并且设置代理后不能访问任何网站。
`2021/02/20 01:43:19 [Info] [313781145] v2ray.com/core/transport/internet/tcp: dialing TCP to tcp:officecdn.microsoft.com:80`
`2021/02/20 01:43:19 [Info] v2ray.com/core/transport/internet/tcp: dialing TCP to tcp:slb-v1.v2clubnode.com:5003`
`2021/02/20 01:43:19 [Info] [313781145] v2ray.com/core/transport/internet/tcp: dialing TCP to tcp:officecdn.microsoft.com:80`
`2021/02/20 01:43:20 [Info] v2ray.com/core/transport/internet/tcp: dialing TCP to tcp:slb-v1.v2clubnode.com:5003`
`2021/02/20 01:43:20 [Info] [313781145] v2ray.com/core/app/proxyman/outbound: failed to process outbound traffic > v2ray.com/core/proxy/freedom: failed to open connection to tcp:officecdn.microsoft.com:80 > v2ray.com/core/common/retry: [dial tcp: lookup officecdn.microsoft.com on [::1]:53: read udp [::1]:38518->[::1]:53: read: connection refused dial tcp: lookup officecdn.microsoft.com on [::1]:53: read udp [::1]:44556->[::1]:53: read: connection refused dial tcp: lookup officecdn.microsoft.com on [::1]:53: read udp [::1]:49558->[::1]:53: read: connection refused dial tcp: lookup officecdn.microsoft.com on [::1]:53: read udp [::1]:46235->[::1]:53: read: connection refused dial tcp: lookup officecdn.microsoft.com on [::1]:53: read udp [::1]:43517->[::1]:53: read: connection refused] > v2ray.com/core/common/retry: all retry attempts failed`
`2021/02/20 01:43:20 [Warning] [313781145] v2ray.com/core/proxy/http: failed to read response from officecdn.microsoft.com > io: read/write on closed pipe`
`2021/02/20 01:43:20 [Info] [313781145] v2ray.com/core/app/proxyman/inbound: connection ends > v2ray.com/core/proxy/http: failed to read http request > EOF`
`2021/02/20 01:43:20 [Info] [930071759] v2ray.com/core/proxy/http: request to Method [CONNECT] Host [api.onedrive.com:443] with URL [//api.onedrive.com:443]`
`2021/02/20 01:43:20 [Info] v2ray.com/core/features/routing/dns: resolve ip for api.onedrive.com > lookup api.onedrive.com on [::1]:53: read udp [::1]:47815->[::1]:53: read: connection refused`
`2021/02/20 01:43:20 [Info] v2ray.com/core/features/routing/dns: resolve ip for api.onedrive.com > lookup api.onedrive.com on [::1]:53: read udp [::1]:43096->[::1]:53: read: connection refused`
`2021/02/20 01:43:20 [Info] v2ray.com/core/features/routing/dns: resolve ip for api.onedrive.com > lookup api.onedrive.com on [::1]:53: read udp [::1]:37539->[::1]:53: read: connection refused`
`2021/02/20 01:43:20 [Info] v2ray.com/core/features/routing/dns: resolve ip for api.onedrive.com > lookup api.onedrive.com on [::1]:53: read udp [::1]:41225->[::1]:53: read: connection refused`
尝试官方编译版本仍然出现此类情况，且注意到程序运行时未出现下列提示：
`[warning] v2ray.com/core/infra/conf: Bootstrap DNS:***`

**Comment by CalmLong** [comment]
重启下手机

或者用 Release 中的版本试试 https://github.com/v2fly/v2ray-core/releases/tag/v4.34.0

**Reply by Waopes** [reply]
重启并使用官网编译版本依然产生上面的问题，感觉固定的dns并没有被应用于程序的DNS解析中。
启动提示如下：
`HWVTR:/data/local/tmp $ ./v2ray`
`V2Ray 4.34.0 (V2Fly, a community-driven edition of V2Ray.) Custom (go1.15.6 android/arm64)`
`A unified platform for anti-censorship.`
`2021/02/20 11:38:16 Using default config:  /data/local/tmp/config.json`
`2021/02/20 11:38:16 [Info] v2ray.com/core/main/jsonem: Reading config: /data/local/tmp/config.json`
`2021/02/20 11:38:18 [Warning] v2ray.com/core: V2Ray 4.34.0 started`

**Reply by Waopes** [reply]
会不会是存在IPv6的原因？6代地址不管怎么都关不掉，所以不知道是不是它造成的。

**Reply by CalmLong** [reply]
APN 设置 - APN 协议 - 仅勾选 IPv4 即可仅使用 IPv4，不过我联通的卡也有 IPv6 的却没这个情况，我的是 Android 10

应该就是代码没生效，但是代码是否生效取决于 Go 语言对操作系统的识别，目前也没好的办法，你也可以这样：

把 [dns_bootstrap_android.go](https://github.com/v2fly/v2ray-core/blob/master/infra/conf/dns_bootstrap_android.go) 的代码复制到 [dns_bootstrap.go](https://github.com/v2fly/v2ray-core/blob/master/infra/conf/dns_bootstrap.go) 覆盖里面的代码(顶部的注释除外)，然后再编译运行

这样就不会根据操作系统来判断了

**Reply by Waopes** [reply]
非常奇怪，根据以上方法编译出来的文件依然不能正确使用。我现在怀疑是不是我编译的问题。。。
过几天看能不能找到Android 10的手机测试一下

**Reply by Waopes** [reply]
尝试按照上述方法编译了一个Windows版本的文件，成功得到了以下提示：
```
V2Ray 4.34.0 (V2Fly, a community-driven edition of V2Ray.) Custom (go1.16 windows/amd64)
A unified platform for anti-censorship.
2021/02/20 22:13:26 Using confdir from env: ……
……
2021/02/20 22:13:28 [Warning] v2ray.com/core/infra/conf: Bootstrap DNS: 8.8.8.8:53
2021/02/20 22:13:29 [Warning] v2ray.com/core: V2Ray 4.34.0 started
```
感觉应该不是编译的问题了，我明天再尝试尝试，总之还是非常感谢！

**Comment by kslr** [comment]
Android 9 不支持的咯

**Reply by Waopes** [reply]
也许吧。。太真实了

**Comment by Waopes** [comment, marked as answer]
在Android 9中测试成功，引起代码不被执行的原因是配置文件中缺少了DNS模块，需要加上
`dns:{}`
本问题的出现应该与系统版本、代码修改无关。
测试似乎表明，在配置文件中添加的DNS解析服务在设置代理情况下可用，但不能监听本地53端口而对外的单独提供DNS解析服务（系统限制）。

**Reply by CalmLong** [reply]
不是吧，现在还有不用内置DNS的吗

还有老哥你这不废话么，要是内置DNS用不了那这样的用法当然没啥意义了呀

大无语事件发生了

**Reply by Waopes** [reply]
。。。我因为之前测试改配置文件，直接把存dns的那个配置文件从文件夹里删掉了。之后用v2rayn自动生成的文件里面也没有这玩意，还是今早偶然发现的。关键是我没想到安卓版本的一定要这东西存在啊。

确实蛮无语的

**Reply by kslr** [reply]
应该不需要也对，不管你写不写配置都会出现bootstrap DNS，你是不是在修改过后的代码上测试

**Reply by CalmLong** [reply]
> 应该不需要也对，不管你写不写配置都会出现bootstrap DNS，你是不是在修改过后的代码上测试

要把这个挪出来才行

 https://github.com/v2fly/v2ray-core/blob/efb963237ca5517d8224733041dabeded6f9b104/infra/conf/dns.go#L161

**Reply by Waopes** [reply]
> 应该不需要也对，不管你写不写配置都会出现bootstrap DNS，你是不是在修改过后的代码上测试

只是修改了代码中的DNS服务器，其他的都没有动。

**Reply by CalmLong** [reply]
> 应该不需要也对，不管你写不写配置都会出现bootstrap DNS，你是不是在修改过后的代码上测试

那要不要把这里该一下，改成不需要配置内置 DNS 的情况下也设置 `8.8.8.8`
